File: Q1_Cpp/src/Product/Clothing/TShirt.cpp
#include "../../../include/Product/Clothing/TShirt.h"
#include <sstream>

// Constructor for TShirt
TShirt::TShirt(std::string id, std::string name, Supplier supplier, int packagingUnits, double price, std::string material, std::string color, std::string size)
    : Clothing(id, name, supplier, packagingUnits, price, material, color), size(size) {
}

// Overriding calculateDiscount method from Product class
double TShirt::calculateDiscount(int quantity, int month) {
    double discountRate = 0.15; // Base discount of 15%
    if (size == "XL") { // Extra discount for XL size
        discountRate += 0.05; // 5% more
    }
    return quantity * price * discountRate; // Calculating total discount
}

// Overriding displayProductInfo method from Product class
std::string TShirt::displayProductInfo() {
    // Returns string representation of TShirt
    std::stringstream ss;
    ss << "TShirt{" <<
       "ID='" << getId() << '\'' <<
       ", Name='" << getName() << '\'' <<
       ", Material='" << getMaterial() << '\'' <<
       ", Color='" << getColor() << '\'' <<
       ", Size='" << size << '\'' <<
       ", Supplier='" << getSupplier().getName() << '\'' <<
       ", Packaging Units=" << getPackagingUnits() <<
       '}';
    return ss.str();
}

// Getter and setter for size
std::string TShirt::getSize() const {
    return size;
}

void TShirt::setSize(const std::string& size) {
    this->size = size;
}

File: Q1_Java/src/main/java/Product/Clothing/TShirt.java
package Product.Clothing; // Package for TShirt

import Product.Supplier;


// TShirt class, extends Clothing
public class TShirt extends Clothing {
    private String size; // Size of the T-shirt (e.g., S, M, L, XL)

    // Constructor for TShirt
    public TShirt(String id, String name, Supplier supplier, int packagingUnits, double price, String material, String color, String size) {
        super(id, name, supplier, packagingUnits, price, material, color); // Calling Clothing's constructor
        this.size = size; // Setting the size
    }

    // Overriding calculateDiscount method
    @Override
    public double calculateDiscount(int quantity, int month) {
        double discountRate = 0.15; // Base discount of 15%
        if ("XL".equals(size)) { // Extra discount for XL size
            discountRate += 0.05; // 5% more
        }
        return quantity * discountRate; // Calculating total discount
    }

    // Overriding displayProductInfo method
    @Override
    public String displayProductInfo() {
        // Returns string representation of TShirt
        return "TShirt{" +
               "ID='" + getId() + '\'' +
               ", Name='" + getName() + '\'' +
               ", Material='" + getMaterial() + '\'' +
               ", Color='" + getColor() + '\'' +
               ", Size='" + size + '\'' +
               ", Supplier='" + getSupplier().getName() + '\'' +
               ", Packaging Units=" + getPackagingUnits() +
               '}';
    }

    // Getter and setter for size
    public String getSize() { return size; }
    public void setSize(String size) { this.size = size; }
}
